### ISSUE TITLE
missing `conv2d` and other convolution operations can just be added straightforwardly?

### ISSUE THREAD

**Issue by silky** [tensorflow/haskell#5]
this project is awesome!

but i note several operations relating to convolutions are missing.

glancing at `tensorflow-ops/src/TensorFlow/Ops.hs` it seems like i can just add in the definitions there in a similar manner; i.e. say:

```
conv2d :: forall a v . TensorType a => Tensor v a -> Tensor v a -> [Int64] -> Build (Tensor Value a)
conv2d input filter strides padding =
    TF.buildOp $ TF.opDef "Conv2D"
            & TF.opAttr "input" .~ input
            & TF.opAttr "filter" .~ filter
            ...
```

right? or is there a bit more to it?

**Comment by blackgnezdo** [comment]
Until we have a good story around Ops, this should be used directly:
https://tensorflow.github.io/haskell/haddock/tensorflow-core-ops-0.1.0.0/TensorFlow-GenOps-Core.html#v:conv2D

**Comment by silky** [comment]
interesting

but is that not missing the strides and padding params? or should i attempt to set those attributes afterwards?

**Comment by blackgnezdo** [comment]
Optional attributes support in generated Ops is another missing piece. I'm afraid you have to resort to manual wrapping like you suggested. I'd copy the code-genned function and update it to your liking...

**Comment by silky** [comment]
no worries @blackgnezdo, thanks for the prompt replies :)

**Comment by fkm3** [comment]
>but is that not missing the strides and padding params? or should i attempt to set those attributes afterwards?

You can set the attributes on the existing `TF.conv2D` like this:

```haskell
TF.conv2D input filter
    & TF.tensorAttr "strides" .~ ([1, 4, 4, 1] :: [Int64])
    & TF.tensorAttr "padding" .~ ("VALID" :: ByteString)
    & TF.tensorAttr "data_format" .~ ("NHWC" :: ByteString)
```

but it is a bit error prone, so it may be better to do what has already been suggested.

**Comment by rikvdkleij** [comment]
@fkm3 We are trying to implement convolutional layer with dilations.

When we run the training we get the following error:
```
018-11-22 17:37:33.926216: W tensorflow/core/framework/op_kernel.cc:1318] OP_REQUIRES failed at conv_grad_filter_ops.cc:352 : Invalid argument: Conv2DCustomBackpropFilter: Size of out_backprop doesn't match computed: actual = 32, computed = 35spatial_dim: 2 input: 36 filter: 2 output: 32 stride: 1 dilation: 1
*** Exception: TensorFlowException TF_INVALID_ARGUMENT "Conv2DCustomBackpropFilter: Size of out_backprop doesn't match computed: actual = 32, computed = 35spatial_dim: 2 input: 36 filter: 2 output: 32 stride: 1 dilation: 1\n\t [[Node: Conv2DBackpropFilter_111 = Conv2DBackpropFilter[T=DT_FLOAT, data_format=\"NHWC\", dilations=[1, 1, 1, 1], padding=\"VALID\", strides=[1, 1, 1, 1], use_cudnn_on_gpu=false, _device=\"/job:localhost/replica:0/task:0/device:CPU:0\"](ExpandDims_25, Shape_110, Reshape_107)]]"
```

The layer works when applying test data on model with untrained parameters. So the problem seems to be the backpropagation.

This is the part which created the convolutional layer:
```
-- Build convolutional layer (1d dilated, residual)
build1dConvOp ::
     Variable Float       -- weights / filters
  -> Variable Float       -- bias
  -> Tensor v Float       -- input node
  -> Int64                -- dilationRate
  -> ByteString           -- padding, i.e. "SAME" or "VALID"
  -> Int64                -- windowSize
  -> Int64                -- nFilters
  -> Tensor Build Float   -- output node
build1dConvOp weights bias inputNode dilationRate padding windowSize nFilters = outputNode
  where
    stridesAttr = opAttr "strides" .~ ([1,1,1,1] :: [Int64])
    dilationAttr = opAttr "dilations" .~ ([1, 1, dilationRate, 1] :: [Int64])
    paddingAttr = opAttr "padding" .~ (padding :: ByteString)
    dataFormatAttr = opAttr "data_format" .~ ("NHWC" :: ByteString)
    attrs = stridesAttr . dilationAttr . paddingAttr . dataFormatAttr
    -- apply expandDims in order to get 2 dimensional data the conv2D can be applied to
    exp_in = expandDims inputNode $ constant [1] [1 :: Int32]
    exp_w = expandDims (readValue weights) $ constant [1] [0 :: Int32]
    conv = conv2D' attrs exp_in exp_w
    squeezed = reshape conv $ vector [-1, windowSize, nFilters]
    outputNode = squeezed `add` readValue bias
```

It seems that our dilations parameter is not propagated to the gradient calculation.

So we suspected that the gradient for `Conv2D` is not correct. But when we add the `dilations` attribute to the gradient we get an error that a dilations rate larger then 1 is not supported. Error:
```
2018-11-22 15:53:29.538087: E tensorflow/core/common_runtime/executor.cc:696] Executor failed to create kernel. Invalid argument: Current libxsmm and customized CPU implementations do not yet support dilation rates larger than 1.
	 [[Node: Conv2DBackpropFilter_111 = Conv2DBackpropFilter[T=DT_FLOAT, data_format="NHWC", dilations=[1, 1, 2, 1], padding="VALID", strides=[1, 1, 1, 1], use_cudnn_on_gpu=false, _device="/job:localhost/replica:0/task:0/device:CPU:0"](ExpandDims_25, Shape_110, Reshape_107)]]
*** Exception: TensorFlowException TF_INVALID_ARGUMENT "Current libxsmm and customized CPU implementations do not yet support dilation rates larger than 1.\n\t [[Node: Conv2DBackpropFilter_111 = Conv2DBackpropFilter[T=DT_FLOAT, data_format=\"NHWC\", dilations=[1, 1, 2, 1], padding=\"VALID\", strides=[1, 1, 1, 1], use_cudnn_on_gpu=false, _device=\"/job:localhost/replica:0/task:0/device:CPU:0\"](ExpandDims_25, Shape_110, Reshape_107)]]"
```

This the python implementation of the `Conv2D` gradient:
```
@ops.RegisterGradient("Conv2D")
def _Conv2DGrad(op, grad):
  dilations = op.get_attr("dilations")
  strides = op.get_attr("strides")
  padding = op.get_attr("padding")
  use_cudnn_on_gpu = op.get_attr("use_cudnn_on_gpu")
  data_format = op.get_attr("data_format")
  shape_0, shape_1 = array_ops.shape_n([op.inputs[0], op.inputs[1]])
  return [
      nn_ops.conv2d_backprop_input(
          shape_0,
          op.inputs[1],
          grad,
          dilations=dilations,
          strides=strides,
          padding=padding,
          use_cudnn_on_gpu=use_cudnn_on_gpu,
          data_format=data_format),
      nn_ops.conv2d_backprop_filter(
          op.inputs[0],
          shape_1,
          grad,
          dilations=dilations,
          strides=strides,
          padding=padding,
          use_cudnn_on_gpu=use_cudnn_on_gpu,
          data_format=data_format)
  ]
```

and the current Haskell implementation differs in the dilations attribute:
```
opGrad "Conv2D" nodeDef [toT -> x, toT -> y] [dz] =
    [ Just $ CoreOps.conv2DBackpropInput'
                ((opAttr "strides" .~ strides)
                    . (opAttr "padding" .~ padding)
        --            . (opAttr "dilations" .~ dilations)
                    . (opAttr "use_cudnn_on_gpu" .~ useCudnnOnGpu)
                    . (opAttr "data_format" .~ dataFormat))
                (shape x) y dz
    , Just $ CoreOps.conv2DBackpropFilter'
                ((opAttr "strides" .~ strides)
        --            . (opAttr "dilations" .~ dilations)
                    . (opAttr "padding" .~ padding)
                    . (opAttr "use_cudnn_on_gpu" .~ useCudnnOnGpu)
                    . (opAttr "data_format" .~ dataFormat))
                x (shape y) dz
    ]
  where
    strides = lookupAttr nodeDef "strides" :: [Int64]
    padding = lookupAttr nodeDef "padding" :: ByteString
   -- dilations = lookupAttr nodeDef "dilations" :: [Int64]
    useCudnnOnGpu = lookupAttr nodeDef "use_cudnn_on_gpu" :: Bool
    dataFormat = lookupAttr nodeDef "data_format" :: ByteString
```

(The comments are our changes to get the gradient working)

@fkm3 Can you help?

**Comment by rikvdkleij** [comment]
An update of trying to implement convolutional layer with dilations.

In the graph of the python implementation we can see that a dilated convolution is translated to a `space_to_batch`, `conv2D`(without dilations) and then `batch_to_space`.

It there no other way then to do this the same way in Haskell? Any example would be really nice!

**Comment by blackgnezdo** [comment]
@rikvdkleij, generally we have very little wrapping around low level (C++) TF ops in Haskell. All the behaviors implemented by the python libraries need to be reimplemented in the Haskell one. Most likely you need to stitch the requisite ops together to create the nicer abstraction that you are seeking. We can either incorporate such work it into this project, or you can provide a separate library layered on top of ours and we can set up some kind of integration testing to keep them in sync.

**Comment by fkm3** [comment]
The "dilations" feature is missing from the gradient implementation because it was added after the conv2d gradient was implemented in haskell.

The error you hit is in libtensorflow. If you wanted to use tensorflow's native dilations feature, you could probably ask about it in the main tensorflow repo. Maybe you need to build libtensorflow in a particular way for it to be supported.
